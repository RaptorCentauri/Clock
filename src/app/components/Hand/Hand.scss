@import '../../styles/animations.scss';
@import '../../styles/handSize.scss';
@import '../../styles/handArrow.scss';


#hour-hand{
  position: relative;
  display: flex;
  justify-content: center;
  align-items: flex-start;

  &:after{
    @include handArrow
  }

  background-color: black;

  @include tick('hour');
  @include handSize('hour');
}


#minute-hand{
  background-color: black;

  position: relative;
  display: flex;
  justify-content: center;
  align-items: flex-start;

  &:after{
    @include handArrow
  }


  @include tick('minute');
  @include handSize('minute');
}

#second-hand{
  background-image: linear-gradient( rgb(203, 6, 0), 75% ,  black);

  display: flex;
  justify-content: center;
  align-items: flex-start;


  @include tick('second');
  @include handSize('second');
}


handRing{
  width: 300px;
  height: 300px;
  border-radius: 50%;
  position: relative;

  grid-area: clockCenter;

  display: flex;
  justify-content: center;
  align-items: center;
}
