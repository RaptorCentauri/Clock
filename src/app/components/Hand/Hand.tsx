import * as React from 'react';
import './Hand.scss';

const Hand = ({handType}) =>{

  //Set up a state for the angle of the hand.
  const [angle, setAngle] = React.useState(0);

  React.useEffect(()=>{
    //Get the current time of the handType passed to the component
    let date = new Date();
    let currentTime = date[`get${handType}s`]();

    //The calculation of the hand angle is dependent on the hand type. Second and Minute hands move 6 degrees per minute. The hour hand moves 0.5 degrees per minute;
    switch(handType){
      case 'Second':
      case 'Minute':
          setAngle(6*currentTime);
      break;

      case 'Hour':
          setAngle(0.5*(60*currentTime+date.getMinutes()));
      break;
    }

});

  //Set the starting angle of the hand.
  const initialAngle = {
    transform: `rotate(${angle}deg)`
  };

  //The handring element is used to setup the startring rotation of the hand. When its not used, the rotation animation will default to 0 degrees for the start position after one complete cycle.
  return(
    <hand-ring style={initialAngle}>
      <clock-hand id={`${handType.toLowerCase()}-hand`}></clock-hand>
    </hand-ring>
  )
}

export default Hand
