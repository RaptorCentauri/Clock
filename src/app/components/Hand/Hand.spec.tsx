import Hand from './Hand';
import * as React from 'react';
import {mount} from 'enzyme';
import * as renderer from 'react-test-renderer';
import { render, fireEvent, getByTestId, cleanup, queryByTestId} from "@testing-library/react";
// import {toBeInTheDocument, toHaveClass, toContainElement} from 'jest-dom'

import 'jest-dom/extend-expect'

afterEach(cleanup)

describe('Hand Component', () => {
  // test('should render when given a handType of Second',()=>{
  //     const handType = 'Second';
  //     const rendered = renderer.create(<Hand handType={handType}/> );
  //    expect(rendered.toJSON()).toMatchSnapshot();
  // });
  //
  // test('should render when given a handType of Minute',()=>{
  //     const handType = 'Minute';
  //     const rendered = renderer.create(<Hand handType={handType}/> );
  //    expect(rendered.toJSON()).toMatchSnapshot();
  // });
  //
  test('should render when given a handType of Hour',()=>{
      const handType = 'Hour';
      const rendered = renderer.create(<Hand handType={handType}/> );
     expect(rendered.toJSON()).toMatchSnapshot();
  });

  test('the hand-ring element to be in the DOM',()=>{
    const handType = 'Hour';
    const { container, asFragment} = render(<Hand handType={handType}/> );
    const handRingElement = container.querySelector('hand-ring');

    //@ts-ignore
    expect(handRingElement).toBeInTheDocument();
  });

  test('the clock-hand element to be in the DOM',()=>{
    const handType = 'Hour';
    const { container} = render(<Hand handType={handType}/> );
    const clockHandElement = container.querySelector('hand-ring').querySelector('clock-hand');
    //
    // //@ts-ignore
    expect(clockHandElement).toBeInTheDocument();

  });



  test('to have an inline transform style',()=>{
    const handType = 'Hour';

    const { container, asFragment} = render(<Hand handType={handType}/> );
    const handRingElement = container.querySelector('hand-ring');

    //@ts-ignore
    expect(handRingElement).toBeInTheDocument();

  // console.log(handRingStyle)

  // console.log(handRingStyle.hasOwnProperty('transform'))

  // console.log(container.firstChild.style)


  // console.log(getByTestId(container), )


    // expect(angleValue.textContent).toBe("0");
//@ts-ignore
  // expect(container).toContainElement('hand-ring');

      // const handType = 'Hour';
      // const rendered = renderer.create(<Hand handType={handType}/> );
     // expect(rendered.toJSON()).toMatchSnapshot();
  });



  // test('should render when given a handType of Hour',()=>{
  //   const handType = 'Hour';
  //   const {container} = render(<Hand handType={handType}/>);
  //
  // });










});
