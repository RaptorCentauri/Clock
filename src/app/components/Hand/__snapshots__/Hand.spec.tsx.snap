// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hand Component should render when given a handType of Hour 1`] = `
<hand-ring
  style={
    Object {
      "transform": "rotate(0deg)",
    }
  }
>
  <clock-hand
    id="hour-hand"
  />
</hand-ring>
`;
